### PR TITLE
refactor: encapsulate wallet's address book access

### PR DESCRIPTION
**Pull request by furszy**
### Context

The wallet's `m_address_book` field is being accessed directly from several places across the sources.

### Problem

Code structure wise, we shouldn't be accessing it directly. It could end up being modified by mistake (from a place that has nothing to do with the wallet like an RPC command or the GUI) and cause a bigger issue: like an address book entry 'purpose' string change, which if done badly (from 'send' to 'receive'), could end up in a user sharing a "receive" address that he/she doesn't own.

### Solution

Encapsulate `m_address_book` access inside the wallet.

-------------------------------------------------------

Extra Note:

This is the initial step towards decoupling the address book functionality from the wallet's sources. In other words, the creation of the `AddressBookManager` (which will be coming in a follow-up PR).

### CODE REVIEW

**Comment by fanquake** [comment]
https://github.com/bitcoin/bitcoin/pull/25337/checks?check_run_id=6844212884:
```bash
  CXX      wallet/rpc/libbitcoin_wallet_a-coins.o
wallet/interfaces.cpp:213:49: error: calling function 'IsMine' requires holding mutex 'm_wallet->cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
            result.emplace_back(dest, m_wallet->IsMine(dest), label, purpose);
                                                ^
```

**Review by hebasto — COMMENTED** [review]
Wrt thread-safety annotations, I'd suggest to:
- rebase this PR to include bitcoin/bitcoin#24931
- ensure that every commit does not introduce new `-Wthread-safety` warnings when compiling with clang
- in the _"wallet: implement ForEachAddrBookEntry method"_ commit (ead512a3d4079b60cc14b88e9076d51f15a6f7e6) apply the following patch:
```diff
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -653,7 +653,8 @@ public:
      * Stops when the provided 'ListAddrBookFunc' returns false.
      */
     using ListAddrBookFunc = std::function<void(const CTxDestination& dest, const std::string& label, const std::string& purpose, bool is_change)>;
-    void ForEachAddrBookEntry(const ListAddrBookFunc& func = [](const CTxDestination&, const std::string&, const std::string&, bool is_change){ return; }) const;
+    void ForEachAddrBookEntry(const ListAddrBookFunc& func = [](const CTxDestination&, const std::string&, const std::string&, bool is_change){ return; }) const
+        EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
     /**
      * Marks all outputs in each one of the destinations dirty, so their cache is
```
- in the _"refactor: use ForEachAddrBookEntry in interfaces::wallet::getAddresses"_ commit (c629bef63135719f4dc6b301788198f557a2ac7d) apply the following patch:
```diff
--- a/src/wallet/interfaces.cpp
+++ b/src/wallet/interfaces.cpp
@@ -208,7 +208,7 @@ public:
     {
         LOCK(m_wallet->cs_wallet);
         std::vector<WalletAddress> result;
-        m_wallet->ForEachAddrBookEntry([&](const CTxDestination& dest, const std::string& label, const std::string& purpose, bool is_change) {
+        m_wallet->ForEachAddrBookEntry([&](const CTxDestination& dest, const std::string& label, const std::string& purpose, bool is_change) EXCLUSIVE_LOCKS_REQUIRED(m_wallet->cs_wallet) {
             if (is_change) return;
             result.emplace_back(dest, m_wallet->IsMine(dest), label, purpose);
         });
```
- and maybe skip the last commit?

**Comment by furszy** [comment]
Thanks for the review hebasto!

Great suggestions 👌🏼 , have applied them all.

> ensure that every commit does not introduce new -Wthread-safety warnings when compiling with clang

Ensured. Have compiled commit by commit with --enable-debug`.

> and maybe skip the last commit?

Done. Agree that it's not related to the scope of this work. Will push it in another PR.

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -648,6 +648,14 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      */
     std::vector<CTxDestination> ListAddrBookAddresses(const std::optional<AddrBookFilter>& filter) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
+    /**
+     * Walk-through the address book entries.
+     * Stops when the provided 'ListAddrBookFunc' returns false.
+     */
+    using ListAddrBookFunc = std::function<void(const CTxDestination& dest, const std::string& label, const std::string& purpose, bool is_change)>;
+    void ForEachAddrBookEntry(const ListAddrBookFunc& func = [](const CTxDestination&, const std::string&, const std::string&, bool is_change){ return; }) const
```
Comment (achow101) [inline src/wallet/wallet.h:661]: In 1db894772fb76d3469410683fc25f1c9721327cd "wallet: implement ForEachAddrBookEntry method"

I don't think it's necessary to have this default parameter for `func`. It isn't used and logically there is no reason why `ForEachAddrBookEntry` would ever be called without a specific `func`.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2362,7 +2362,7 @@ std::vector<CTxDestination> CWallet::ListAddrBookAddresses(const std::optional<A
     AssertLockHeld(cs_wallet);
     std::vector<CTxDestination> result;
     AddrBookFilter filter = _filter ? *_filter : AddrBookFilter();
-    ForEachAddrBookEntry([&result, &filter](const CTxDestination& dest, const std::string& label, const std::string& purpose, bool is_change){
+    ForEachAddrBookEntry([&result, &filter](const CTxDestination& dest, const std::string& label, const std::string& purpose, bool is_change) {
```
Comment (achow101) [inline src/wallet/wallet.cpp:2365]: In 38b871da82cfc7fe74cc7f844c8c6636a0a0234c "refactor: RPC 'listlabels', encapsulate 'CWallet::ListAddrBookLabels' functionality"

nit: This change should be in 1db894772fb76d3469410683fc25f1c9721327cd "wallet: implement ForEachAddrBookEntry method"

File: src/wallet/wallet.h (hunk 1 above)
Comment (furszy) [reply src/wallet/wallet.h:661]: yeah, good catch

**Comment by furszy** [comment]
Thanks for the review achow101

Updated per feedback. Removed `ForEachAddrBookEntry` default parameter.

**Comment by achow101** [comment]
ACK 21c6342555f4e037f51644c4b9e77caa0e3c5fa3

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -635,7 +635,30 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     std::optional<int64_t> GetOldestKeyPoolTime() const;
 
-    std::set<CTxDestination> GetLabelAddresses(const std::string& label) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    // Filter struct for 'ListAddrBookAddresses'
+    struct AddrBookFilter {
+        // Fetch addresses with the provided label
+        std::optional<std::string> m_op_label{std::nullopt};
+        // Always filter change addresses by default
+        bool filter_change{true};
```
Comment (theStack) [inline src/wallet/wallet.h:643]: nit: The usage of the verb "filter" is potentially confusing here, as it's not clear whether it means inclusion or exclusion; if the filter structure has `m_op_label` set, it means to only end up with addresses of that label, but if `filter_change` is true, all change addresses are ignored (i.e. the opposite logic). Probably use something like

```suggestion
        // Never include change addresses by default
        bool ignore_change{true};
```

**Review by theStack — COMMENTED** [review]
Concept ACK

File: src/wallet/wallet.h (hunk 3 above)
Comment (furszy) [reply src/wallet/wallet.h]: sure, updated.

**Comment by furszy** [comment]
Thanks for the feedback!

Only modification is a variable rename `filter_change` -> `ignore_change`.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2348,21 +2348,45 @@ void CWallet::MarkDestinationsDirty(const std::set<CTxDestination>& destinations
...
+    AssertLockHeld(cs_wallet);
+    std::set<std::string> label_set;
+    ForEachAddrBookEntry([&](const CTxDestination& _dest, const std::string& _label,
+                             const std::string& _purpose, bool _is_change) {
+        if (_is_change) return;
+        if (purpose.empty() || _purpose == purpose) {
+            label_set.insert(_label);
+        }
+    });
+   return label_set;
```
Comment (theStack) [inline src/wallet/wallet.cpp:2387]: nit: missing space
```suggestion
    return label_set;
```

File: src/wallet/rpc/transactions.cpp [hunk 5]
```diff
@@ -138,70 +134,53 @@ static UniValue ListReceived(const CWallet& wallet, const UniValue& params, cons
...
-        }
-    }
-
-    for (auto item_it = start; item_it != end; ++item_it)
-    {
-        if (item_it->second.IsChange()) continue;
-        const CTxDestination& address = item_it->first;
-        const std::string& label = item_it->second.GetLabel();
+    const auto& func = [&](const CTxDestination& address, const std::string& label, const std::string& purpose, bool is_change) {
         auto it = mapTally.find(address);
```
Comment (theStack) [inline src/wallet/rpc/transactions.cpp:138]: To match the previous behaviour of the iteration loop, I think change addresses should be ignored?
```suggestion
        if (is_change) return;
        auto it = mapTally.find(address);
```

**Review by theStack — COMMENTED** [review]
Almost-ACK, changes look good to me. Just two suggestions below, one nit, one potential behaviour change.

**Review by theStack — APPROVED** [review]
ACK d69045e291e32e02d105d1b5ff1c8b86db0ae69e ✅

**Comment by achow101** [comment]
ACK d69045e291e32e02d105d1b5ff1c8b86db0ae69e

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25337/commits/d69045e291e32e02d105d1b5ff1c8b86db0ae69e

It's an improvement in code organization, as ideally only `CWallet` would be aware of `m_address_book`.
This field is still called in `walletdb.cpp` and `wallettool.cpp`, but may eventually be handled in another PR.

File: src/interfaces/wallet.h [hunk 6]
```diff
@@ -114,7 +114,7 @@ class Wallet
         std::string* purpose) = 0;
 
     //! Get wallet address list.
-    virtual std::vector<WalletAddress> getAddresses() = 0;
+    virtual std::vector<WalletAddress> getAddresses() const = 0;
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/interfaces/wallet.h:117]: k 👍🏼, got it.

**Pull request merged**